File: src/bin/day16p1.rs
#![feature(once_cell)]
#![feature(saturating_int_impl)]
#![feature(test)]

extern crate test;

use std::num::Saturating;

use aoc2022::lazily;
use regex::Regex;
use rustc_hash::FxHashMap;
use smallvec::SmallVec;

fn main() {
    let input = std::io::read_to_string(std::io::stdin()).unwrap();
    let input = parse(&input);
    let a = solve(&input);
    println!("{a}");
}

fn brute_force(
    t: usize,
    i: usize,
    dist: &[Vec<Saturating<usize>>],
    seen: usize,
    pressurised_valves: &[usize],
    nodes: &[Node],
) -> usize {
    let mut best = 0;

    for &j in pressurised_valves {
        if (seen & (1 << j)) == 0 {
            let new_t = t + dist[i][j].0 + 1;
            if new_t <= 30 {
                let a = brute_force(new_t, j, dist, seen | (1 << j), pressurised_valves, nodes)
                    + (30 - new_t) * nodes[j].pressure;
                if a > best {
                    best = a;
                }
            }
        }
    }

    best
}

fn solve(input: &Input) -> usize {
    let n = input.nodes.len();

    let mut dist = vec![vec![Saturating(usize::MAX); n]; n];
    for (index, Node { neighbours, .. }) in input.nodes.iter().enumerate() {
        for &i in neighbours {
            dist[index][i] = Saturating(1);
        }
    }

    for k in 0..n {
        for i in 0..n {
            for j in 0..n {
                if dist[i][j] > dist[i][k] + dist[k][j] {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    let mut pressurised_valves = vec![];
    for (index, Node { pressure, .. }) in input.nodes.iter().enumerate() {
        if pressure > &0 {
            pressurised_valves.push(index);
        }
    }

    brute_force(0, input.root, &dist, 0, &pressurised_valves, &input.nodes)
}

#[derive(Clone, PartialEq, Eq, Debug)]
struct Node {
    neighbours: SmallVec<[usize; 2]>,
    pressure: usize,
}

#[derive(Clone, PartialEq, Eq, Debug)]
struct Input {
    nodes: Vec<Node>,
    root: usize,
}

fn parse(input: &str) -> Input {
    let regex = lazily!(Regex::new(r"[A-Z]{2}").unwrap());
    let digits = lazily!(Regex::new(r"\d+").unwrap());

    let mut name_to_index: FxHashMap<&str, usize> = FxHashMap::default();
    let mut nodes_in_order = vec![];
    for (index, line) in input.lines().enumerate() {
        let pressure = digits.find(line).unwrap().as_str().parse().unwrap();

        let mut names = line.matches(regex);
        let from = names.next().unwrap();

        name_to_index.insert(from, index);
        nodes_in_order.push((names, pressure));
    }

    let mut nodes = vec![];
    for (neighbours, pressure) in nodes_in_order {
        nodes.push(Node {
            neighbours: neighbours.map(|node| name_to_index[node]).collect(),
            pressure,
        });
    }

    Input {
        nodes,
        root: name_to_index["AA"],
    }
}

#[bench]
fn bench(bencher: &mut test::Bencher) {
    let input = include_str!("../../inputs/day16");
    let input = parse(input);

    bencher.iter(|| {
        test::black_box(solve(&input))
    });
}

File: day16p1.py
import re
import sys
from typing import Dict, List, Tuple, Iterator


class Node:
    neighbours: List[int]
    pressure: int

    def __init__(self, neighbours: List[int], pressure: int) -> None:
        self.neighbours = neighbours
        self.pressure = pressure


def parse(input: str):
    vault_regex = re.compile(r"[A-Z]{2}")
    number_regex = re.compile(r"\d+")

    name_to_index: Dict[str, int] = dict()
    nodes_in_order: List[Tuple[List[str], int]] = []
    index = 0
    for line in input.splitlines():
        pressure = int(number_regex.findall(line)[0])

        names = vault_regex.findall(line)
        here = names[0]

        name_to_index[here] = index
        nodes_in_order.append((names[1:], pressure))

        index += 1

    nodes = []
    for neighbours, pressure in nodes_in_order:
        nodes.append(Node(
            neighbours=[name_to_index[name] for name in neighbours],
            pressure=pressure,
        ))

    return nodes, name_to_index["AA"]


def solve(nodes: List[Node], root: int) -> int:
    n = len(nodes)

    dist = [[float("inf")] * n for _ in range(n)]
    for i, node in enumerate(nodes):
        for neighbour in node.neighbours:
            dist[i][neighbour] = 1
        dist[i][i] = 0

    # Floyd-Warshall
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][j] > dist[i][k] + dist[k][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    pressurised_valves = [
        i
        for i, node in enumerate(nodes)
        if node.pressure > 0
    ]

    n_pressurised = len(pressurised_valves)

    def brute_force(t: int, i: int, seen: int) -> int:
        best = 0

        for seen_j, j in enumerate(pressurised_valves):
            if not (seen & (1 << seen_j)):
                new_t = t + dist[i][j] + 1
                if new_t <= 30:
                    a = brute_force(new_t, j, seen | (1 << seen_j)) + \
                        (30 - new_t) * nodes[j].pressure

                    if a > best:
                        best = a

        return best

    return brute_force(0, root, 0)


def main():
    input = sys.stdin.read()
    nodes, root = parse(input)
    part1 = solve(nodes, root)
    print(part1)


main()
